### ISSUE TITLE
Fallback to default language 

### ISSUE THREAD

**Issue by jonandersen** [SwiftGen/SwiftGen#381]
I'm currently using swiftgen to generate Strings for my app. However some of my translations aren't complete. The current behavior with SwiftGen (and Apple) is that the key is returned if it can't be found.

I have this following code to work around this (if key not found try the English translations)

``` swift
extension L10n {
  private static func tr(_ table: String, _ key: String, _ args: CVarArg...) -> String {
    return localizeStringAndFallbackToEn(table, key, args)
  }
}

func localizeStringAndFallbackToEn(_ table: String, _ key: String, _ args: CVarArg...) -> String {
    var format = NSLocalizedString(key, tableName: table, bundle: Bundle(for: BundleToken.self), comment: "")
    let value = String(format: format, locale: Locale.current, arguments: args)
    if value != key || NSLocale.preferredLanguages.first == "en" {
        return String(format: format, locale: Locale.current, arguments: args)
    }
    
    // Fall back to en
    guard
        let path = Bundle.main.path(forResource: "en", ofType: "lproj"),
        let bundle = Bundle(path: path)
        else { return value }
    format =  NSLocalizedString(key, bundle: bundle, comment: "")
    return String(format: format, locale: Locale.current, arguments: args)
}
```

Is it possible to add a behavior like this into SwiftGen? Maybe an optional flag to generate a fallback to a language.

**Comment by djbe** [comment]
I'm not sure that something many people would ever need. Most people use something like the XLIFF export/import flow, and any (decent) XLIFF editor will point out where there are any missing translations.

Also, you're defaulting here to "en" as a fallback language. Not everyone has english as a development/base language, or they might be using a variation of english (`en-GB` for example). How would SwiftGen even know what the development language is?

Side note: Falling back to another language isn't something you'd actually want for your users TBH.

**Comment by AliSoftware** [comment]
IIRC you shouldn't even have to implement the fallback yourself, as if the key isn't find in the current locale the user is using, is will first try to fallback to the non-region-specific locale (e.g. if no `fr-CA` it will fall back to `fr`), then will fall back to Base, then the development region, and only if all fallbacks fail it will return the string passed as the `value` parameter to `NSLocalizableStringFromTable(key, table, value, comment, …)`, which defaults to the key itself when you use the common convenience function `NSLocalizedString` paying only the key and value.

**Comment by AliSoftware** [comment]
Wait or maybe it's just to determine the lproj folder to use at app startup? 🤔https://developer.apple.com/library/content/qa/qa1828/_index.html

**Comment by jonandersen** [comment]
I just found the following option in the service I'm using. That will actually solve the problem for me :)

Thanks!

<img width="420" alt="screen shot 2018-02-04 at 11 29 00 pm" src="https://user-images.githubusercontent.com/1019479/35788472-38f750c4-0a03-11e8-90c6-ce2e47f61f15.png">

**Comment by jonandersen** [comment]
@AliSoftware Yea I think it's only to pick that. If the key doesn't exist in there it will return the key back. Maybe it's different for en-GB (never tested that). But for instance I was missing a few words in my `zh-HANS` and it didn't default to the words I have in `Base` or `en` lproj.

**Comment by chrisjrex** [comment]
@djbe / @AliSoftware 

It would be great to look into this issue again. To state the issue more clearly, ill take a paragraph from this article that explains it well.

[Article link](https://samwize.com/2018/01/23/localization-fall-back-to-base-language/)

<img width="700" alt="CleanShot 2022-05-09 at 14 21 57@2x" src="https://user-images.githubusercontent.com/4457170/167419193-dca151a0-fdea-4aa6-b025-61749645c837.png">

> **Fallback 3: Unsupported Phrase**
Let’s take the example where a user prefers zh-Hans.
>
>What happens when iOS try to get the NSLocalizedString of `“another-title”`?
Remember, zh-Hans is not fully translated. It does not have `“another-title”`. What do you think will happen?
>
>1) Fallback to development language en, or
>2) return `“another-title”`
>
>Many developers think it is (1). Unfortunately, it is (2), with that ugly key name!

---------------------

Xcode does not handle fallback for partially translated .strings files, only when the file is missing.

Here is an updated version of the code by @jonandersen, that handles the issue @djbe was concerned about with the development language not being `en`

```
func localizeStringAndFallbackToEn(_ table: String, _ key: String, _ args: CVarArg...) -> String {
    let bundle = BundleToken.bundle
    let format = NSLocalizedString(key, tableName: table, bundle: bundle, comment: "")
    let string = String(format: format, locale: Locale.current, arguments: args)

    // If string key does not find a match within the Locale.current file. Look up the fallback string.
    // The edge case of this is that if a key and value are expected to match: "Foo" = "Foo", it will trigger the fallback logic.
    if string == key,
       let developmentRegion = bundle.infoDictionary?["CFBundleDevelopmentRegion"] as? String,
       let path = bundle.path(forResource: developmentRegion, ofType: "lproj"),
       let bundle = Bundle(path: path)
    {
        let format = NSLocalizedString(key, bundle: bundle, comment: "")
        return String(format: format, locale: Locale(identifier: developmentRegion), arguments: args)
    } else {
        return string
    }
}
```

-------------

I have tested this solution, and created a custom stencil implementing this logic. However, this complicates maintenance of the repo when SwiftGen releases updated stencils for new Swift versions.

I also think it would be a useful solution to others. Where showing fallback text is likely a better "bug" than displaying a key

**Comment by chrisjrex** [comment]
opened a new issue since this one was closed - #937

**Comment by djbe** [comment]
Implemented by SwiftGen 6.6.0, which has been released.
